demos/pkey/EVP_PKEY_RSA_keygen.c

```diff
@@ -0,0 +1,279 @@
+
+    /* Create context using RSA algorithm. "RSA-PSS" could also be used here. */
+    genctx = EVP_PKEY_CTX_new_from_name(libctx, "RSA", propq);
+    if (genctx == NULL) {
+        fprintf(stderr, "EVP_PKEY_CTX_new_from_name() failed\n");
+        goto cleanup;
+    }
+
+    /* Initialize context for key generation purposes. */
+    if (EVP_PKEY_keygen_init(genctx) <= 0) {
+        fprintf(stderr, "EVP_PKEY_keygen_init() failed\n");
+        goto cleanup;
+    }
+
+    /*
+     * Here we set the number of bits to use in the RSA key.
+     * See comment at top of file for information on appropriate values.
+     */
+    if (EVP_PKEY_CTX_set_rsa_keygen_bits(genctx, bits) <= 0) {
+        fprintf(stderr, "EVP_PKEY_CTX_set_rsa_keygen_bits() failed\n");
+        goto cleanup;
+    }
+
+    /*
+     * It is possible to create an RSA key using more than two primes.
+     * Do not do this unless you know why you need this.
+     * You ordinarily do not need to specify this, as the default is two.
+     *
+     * Both of these parameters can also be set via EVP_PKEY_CTX_set_params, but
+     * these functions provide a more concise way to do so.
+     */
+    if (EVP_PKEY_CTX_set_rsa_keygen_primes(genctx, primes) <= 0) {
+        fprintf(stderr, "EVP_PKEY_CTX_set_rsa_keygen_primes() failed\n");
+        goto cleanup;
+    }
+
+    /*
+     * Generating an RSA key with a number of bits large enough to be secure for
+     * modern applications can take a fairly substantial amount of time (e.g.
+     * one second). If you require fast key generation, consider using an EC key
+     * instead.
+     *
+     * If you require progress information during the key generation process,
+     * you can set a progress callback using EVP_PKEY_set_cb; see the example in
+     * EVP_PKEY_generate(3).
+     */
+    fprintf(stderr, "Generating RSA key, this may take some time...\n");
+    if (EVP_PKEY_generate(genctx, &pkey) <= 0) {
+        fprintf(stderr, "EVP_PKEY_generate() failed\n");
+        goto cleanup;
+    }
+
+    /* pkey is now set to an object representing the generated key pair. */
+
+cleanup:
+    EVP_PKEY_CTX_free(genctx);
+    return pkey;
+}
+
+/*
+ * Generates an RSA public-private key pair and returns it.
+ * The number of bits is specified by the bits argument.
+ *
+ * This uses a more concise way of generating an RSA key, which is suitable for
+ * simple cases. It is shown for demonstration but is not used by the code
+ * below.
+ */
+#if 0
+static EVP_PKEY *generate_rsa_key_short(OSSL_LIB_CTX *libctx, unsigned int bits)
+{
+    EVP_PKEY *pkey = NULL;
+
+    fprintf(stderr, "Generating RSA key, this may take some time...\n");
+    pkey = EVP_PKEY_Q_keygen(libctx, propq, "RSA", (size_t)bits);
+
+    if (pkey == NULL)
+        fprintf(stderr, "EVP_PKEY_Q_keygen() failed\n");
+
+    return pkey;
+}
+#endif
+
+/*
+ * Prints information on an EVP_PKEY object representing an RSA key pair.
+ */
+static int dump_key(const EVP_PKEY *pkey)
+{
```

Comment:
EVP_PKEY_print_private() is an easier way of doing this...

Reply:
That's true, though the idea was really to show how to retrieve the individual components of the key. Printing them is mostly an excuse for that.

Reply:
The EC one is doing the same,so it is consistent with that one..